src/source-map.ts

```diff
@@ -0,0 +1,192 @@
+
+function offsetSourceMap(
+    sourcemap: RawSourceMap, lineOffset: number, firstLineCharOffset: number) {
+  const consumer = new SourceMapConsumer(sourcemap);
+  const generator = new SourceMapGenerator();
+
+  consumer.eachMapping(mapping => {
+    const newMapping: any = {
+      source: mapping.source,
+      original: {line: mapping.originalLine, column: mapping.originalColumn},
+      generated: {
+        line: mapping.generatedLine + lineOffset,
+        column: mapping.generatedColumn +
+            (mapping.generatedLine === 1 ? firstLineCharOffset : 0)
+      }
+    };
+
+    if (mapping.name) {
+      newMapping.name = mapping.name;
+    }
+
+    generator.addMapping(newMapping);
+  });
+
+  return generator.toJSON();
+}
+
+export async function getExistingSourcemap(
+    analyzer: AnalysisContext, sourceUrl: string, sourceContent: string) {
+  const dataUrlPrefix = 'data:application/json;';
+  const sourceMappingUrlParts = sourceMappingUrlExpr.exec(sourceContent);
+  if (sourceMappingUrlParts === null) {
+    return null;
+  }
+
+  let sourcemap: RawSourceMap;
+  let mapUrl = sourceUrl;
+  if (sourceMappingUrlParts[1].substr(0, dataUrlPrefix.length) ===
+      dataUrlPrefix) {
+    sourcemap = JSON.parse(Buffer.from(sourceMappingUrlParts[1], 'base64')
+                               .toString('utf8')) as RawSourceMap;
+  } else {
+    mapUrl = urlLib.resolve(sourceUrl, sourceMappingUrlParts[1].trim());
+    sourcemap = JSON.parse(await analyzer.load(mapUrl)) as RawSourceMap;
+  }
+
+  // Rewrite the sources array to be relative to the current URL
+  if (sourcemap.sources) {
+    sourcemap.sources =
+        sourcemap.sources.map(source => urlLib.resolve(mapUrl, source));
+  }
+  return sourcemap;
+}
+
+export async function addOrUpdateSourcemapComment(
+    analyzer: AnalysisContext,
+    sourceUrl: string,
+    sourceContent: string,
+    lineOffset: number = 0,
+    firstLineCharOffset: number = 0) {
+  let sourcemap =
+      await getExistingSourcemap(analyzer, sourceUrl, sourceContent);
+
+  let hasExisting = true;
+  if (sourcemap === null) {
+    hasExisting = false;
+    sourcemap = createJsIdentitySourcemap(
+        sourceUrl, sourceContent, lineOffset, firstLineCharOffset);
+  }
+
+  if (sourcemap === null) {
+    return sourceContent;
+  }
+
+  let updatedSourcemapComment = inlineSourcemapPrefix +
+      Buffer.from(JSON.stringify(sourcemap), 'utf8').toString('base64') + '\n';
+  if (hasExisting) {
+    return sourceContent.replace(sourceMappingUrlExpr, updatedSourcemapComment);
+  } else {
+    if (sourceContent.length > 0 &&
+        sourceContent[sourceContent.length - 1] === '\n') {
+      updatedSourcemapComment = updatedSourcemapComment.substr(1);
+    }
+
+    return sourceContent + updatedSourcemapComment;
+  }
+}
+
+
+/**
+ * Update the source maps in inline script elements to reflect their relative
+ * line offset in the bundle.
```

Comment:
Not quite. As this occurs during the import process, the final position of any token in the HTML document is unknown at this point. Instead, the mappings reflect the relative position within the script tag. It's not a correct source mapping yet - but an intermediate state.